core/src/main/java/com/radixdlt/genesis/olympia/OlympiaGenesisService.java

```diff
@@ -0,0 +1,260 @@
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.genesis.olympia;
+
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+import com.google.common.hash.HashCode;
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.inject.Inject;
+import com.radixdlt.crypto.ECDSASecp256k1Signature;
+import com.radixdlt.crypto.HashUtils;
+import com.radixdlt.genesis.GenesisData;
+import com.radixdlt.genesis.olympia.OlympiaEndStateApiClient.OlympiaEndStateResponse;
+import com.radixdlt.genesis.olympia.state.OlympiaStateIRDeserializer;
+import com.radixdlt.genesis.olympia.state.OlympiaStateIRSerializationException;
+import com.radixdlt.networks.Network;
+import com.radixdlt.utils.Bytes;
+import com.radixdlt.utils.ThreadFactories;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.xerial.snappy.Snappy;
+
+@SuppressWarnings({"UnstableApiUsage", "OptionalUsedAsFieldOrParameterType"})
+public final class OlympiaGenesisService {
+  private static final Logger log = LogManager.getLogger();
+
+  // A placeholder hash included in the "not ready" response
+  // used to be able to verify that the configured Olympia node public key
+  // is correct, even before the network has shutdown. Used as an early
+  // feedback mechanism in case of a misconfiguration. Must be the same
+  // on Olympia and Babylon nodes.
+  private static final HashCode PLACEHOLDER_HASH_FOR_NOT_READY_RESPONSE =
+      HashUtils.sha256Twice(new byte[] {1});
+
+  // Polling interval when connection to Olympia node succeeds
+  private static final long POLL_INTERVAL_AFTER_NOT_READY_MS = 1000;
+
+  // Polling interval when there was an error
+  private static final long POLL_INTERVAL_AFTER_ERROR_MS = 10000;
+
+  private final Network network;
+  private final OlympiaGenesisConfig olympiaGenesisConfig;
+  private final OlympiaEndStateApiClient olympiaEndStateApiClient;
+  private final RateLimiter notReadyLogRateLimiter = RateLimiter.create(0.03);
+
+  private Optional<ScheduledExecutorService> executor = Optional.empty();
+
+  @Inject
+  public OlympiaGenesisService(
+      Network network,
+      OlympiaGenesisConfig olympiaGenesisConfig,
+      OlympiaEndStateApiClient olympiaEndStateApiClient) {
+    this.network = network;
+    this.olympiaGenesisConfig = olympiaGenesisConfig;
+    this.olympiaEndStateApiClient = olympiaEndStateApiClient;
+  }
+
+  public void start(
```

Comment:
I see the `.start()/.shutdown()`, and an internal executor, etc. and it all looks like some async logic... but conceptually (and also in "runtime practice") this is all a part of a very synchronous process (i.e. the app boot-up won't proceed until one of the callbacks is called), right?

In such case, wouldn't we be better off with a simple synchronous code, making calls in a loop, with good-old `.sleep()` in-between?

Reply:
I think it's better not trying to hide its async nature. It works quite nice now with the new redesigned node bootstrapper, wdyt?

Reply:
The new bootstrapper really is an improvement 👍 
Can we propagate the `Future<>` idea that one call further?
(I mean, return `Future<GenesisData> start()` here? Does it cause problems? Certainly it wouldn't hide the async nature of this service [boost it, if anything]).